internal/zinc-core/src/main/scala/sbt/internal/inc/Stamp.scala

```diff
@@ -58,10 +59,10 @@ trait WithPattern { protected def Pattern: Regex }
 import java.lang.{ Long => BoxedLong }
 
 /** Define the hash of the file contents. It's a typical stamp for compilation sources. */
-final class Hash(val value: Array[Byte]) extends StampBase {
-  val hexHash = IOHash.toHex(value)
+final class Hash private[inc] (val hexHash: String) extends StampBase {
```

Comment:
> What does "The validation is in the types" mean? I'd just like it be impossible to create illegal Hash objects. Currently it's trivial.
  
What I mean is that it's not possible to prevent the creation of illegal hash objects because `Stamp` needs access to a constructor that takes a string. Whether that is provided by a constructor or a method in the companion object produces the same result: it has to be accessible from `Stamp`, which disqualifies it from being defined as `private[this]`. The only way this can be supported is by wrapping all the definitions inside an object or a package, and then fixing all the imports. Too big of a change to do it here. I'm happy to do open a ticket to do it.
  
All in all, with the current organisation of the code, `private[this]` cannot be used and that's why we use `private[inc]`, which at least gives us something.